File: tcb_manufacturing_customizations/tcb_manufacturing_customizations/page/sales_flow_dashboard/sales_flow_dashboard.js
frappe.pages['sales-flow-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Sales Flow Dashboard',
        single_column: true
    });

    // Load page content
    $(frappe.render_template('sales_flow_dashboard')).appendTo(page.body);

    // Initialize the dashboard
    new SalesFlowDashboard(page);
};

class SalesFlowDashboard {
    constructor(page) {
        this.page = page;
        this.$wrapper = $(page.body);
        this.filters = {};

        this.init();
    }

    init() {
        this.setup_filters();
        this.bind_events();
        this.load_data();
    }

    setup_filters() {
        const me = this;

        // Company filter
        this.company_field = frappe.ui.form.make_control({
            df: {
                fieldtype: 'Link',
                options: 'Company',
                fieldname: 'company',
                placeholder: 'Select Company',
                default: frappe.defaults.get_user_default('Company')
            },
            parent: this.$wrapper.find('#company-filter'),
            render_input: true
        });
        this.company_field.set_value(frappe.defaults.get_user_default('Company'));

        // From Date filter
        this.from_date_field = frappe.ui.form.make_control({
            df: {
                fieldtype: 'Date',
                fieldname: 'from_date',
                placeholder: 'From Date',
                default: frappe.datetime.month_start()
            },
            parent: this.$wrapper.find('#from-date-filter'),
            render_input: true
        });
        this.from_date_field.set_value(frappe.datetime.month_start());

        // To Date filter
        this.to_date_field = frappe.ui.form.make_control({
            df: {
                fieldtype: 'Date',
                fieldname: 'to_date',
                placeholder: 'To Date',
                default: frappe.datetime.month_end()
            },
            parent: this.$wrapper.find('#to-date-filter'),
            render_input: true
        });
        this.to_date_field.set_value(frappe.datetime.month_end());
    }

    bind_events() {
        const me = this;

        // Refresh button
        this.$wrapper.find('#refresh-btn').on('click', () => {
            me.load_data();
        });

        // Card click to navigate with filters
        this.$wrapper.on('click', '.flow-card', function() {
            const doctype = $(this).data('doctype');
            if (doctype) {
                const filters = me.get_list_filters(doctype);
                frappe.set_route('List', doctype, filters);
            }
        });

        // Pending item click
        this.$wrapper.on('click', '.pending-item, .recent-item', function() {
            const doctype = $(this).data('doctype');
            const name = $(this).data('name');
            if (doctype && name) {
                frappe.set_route('Form', doctype, name);
            }
        });

        // Customer item click
        this.$wrapper.on('click', '.customer-item', function() {
            const customer = $(this).data('customer');
            if (customer) {
                frappe.set_route('Form', 'Customer', customer);
            }
        });
    }

    get_list_filters(doctype) {
        const from_date = this.from_date_field.get_value();
        const to_date = this.to_date_field.get_value();
        const company = this.company_field.get_value();

        // Different doctypes have different date field names
        const date_field_map = {
            'Quotation': 'transaction_date',
            'Sales Order': 'transaction_date',
            'Release Order': 'posting_date',
            'Delivery Note': 'posting_date'
        };

        const date_field = date_field_map[doctype] || 'creation';
        const filters = {};

        if (from_date) {
            filters[date_field] = ['>=', from_date];
        }
        if (to_date) {
            filters[date_field] = from_date
                ? ['between', [from_date, to_date]]
                : ['<=', to_date];
        }
        if (company) {
            filters['company'] = company;
        }

        return filters;
    }

    show_loading() {
        this.$wrapper.find('#loading-overlay').addClass('active');
    }

    hide_loading() {
        this.$wrapper.find('#loading-overlay').removeClass('active');
    }

    load_data() {
        const me = this;
        me.show_loading();

        frappe.call({
            method: 'tcb_manufacturing_customizations.tcb_manufacturing_customizations.page.sales_flow_dashboard.sales_flow_dashboard.get_dashboard_data',
            args: {
                from_date: me.from_date_field.get_value(),
                to_date: me.to_date_field.get_value(),
                company: me.company_field.get_value()
            },
            callback: function(r) {
                me.hide_loading();
                if (r.message) {
                    me.render_dashboard(r.message);
                }
            },
            error: function() {
                me.hide_loading();
                frappe.msgprint(__('Error loading dashboard data'));
            }
        });
    }

    render_dashboard(data) {
        this.render_summary_cards(data.summary_cards);
        this.render_funnel_chart(data.flow_funnel);
        this.render_trend_chart(data.daily_trend);
        this.render_bales_stats(data.bales_stats);
        this.render_batch_stats(data.batch_consumption);
        this.render_ro_stats(data.release_order_stats);
        this.render_top_customers(data.top_customers);
        this.render_pending_documents(data.pending_documents);
    }

    render_summary_cards(data) {
        // Quotation
        this.animate_number('#quotation-count', data.quotation.count);
        this.$wrapper.find('#quotation-value').text(this.format_currency(data.quotation.value));
        this.$wrapper.find('#quotation-open').text(data.quotation.open);
        this.$wrapper.find('#quotation-ordered').text(data.quotation.ordered);

        // Sales Order
        this.animate_number('#so-count', data.sales_order.count);
        this.$wrapper.find('#so-value').text(this.format_currency(data.sales_order.value));
        this.$wrapper.find('#so-pending').text(data.sales_order.pending_delivery);
        this.$wrapper.find('#so-qty').text(this.format_number(data.sales_order.qty));

        // Release Order
        this.animate_number('#ro-count', data.release_order.count);
        this.$wrapper.find('#ro-qty').text(this.format_number(data.release_order.qty) + ' Qty Released');
        this.$wrapper.find('#ro-draft').text(data.release_order.draft);
        this.$wrapper.find('#ro-submitted').text(data.release_order.submitted);

        // Delivery Note
        this.animate_number('#dn-count', data.delivery_note.count);
        this.$wrapper.find('#dn-value').text(this.format_currency(data.delivery_note.value));
        this.$wrapper.find('#dn-qty').text(this.format_number(data.delivery_note.qty));
    }

    render_funnel_chart(data) {
        const container = this.$wrapper.find('#funnel-chart');
        container.empty();

        if (!data || data.length === 0) {
            container.html('<div class="empty-state">No data available</div>');
            return;
        }

        const maxCount = Math.max(...data.map(d => d.count));

        data.forEach((item, index) => {
            const width = maxCount > 0 ? Math.max((item.count / maxCount) * 100, 10) : 10;
            const html = `
                <div class="funnel-item fade-in fade-in-delay-${index + 1}">
                    <span class="funnel-label">${item.stage}</span>
                    <div class="funnel-bar-container">
                        <div class="funnel-bar" style="width: ${width}%; background: ${item.color};">
                            ${this.format_number(item.count)}
                        </div>
                    </div>
                </div>
            `;
            container.append(html);
        });
    }

    render_trend_chart(data) {
        const container = this.$wrapper.find('#trend-chart');

        // Prepare datasets
        const dates = new Set();
        data.sales_orders.forEach(d => dates.add(d.date));
        data.delivery_notes.forEach(d => dates.add(d.date));
        data.release_orders.forEach(d => dates.add(d.date));

        const sortedDates = Array.from(dates).sort();

        // Create data maps
        const soMap = {};
        data.sales_orders.forEach(d => soMap[d.date] = d.count);

        const dnMap = {};
        data.delivery_notes.forEach(d => dnMap[d.date] = d.count);

        const roMap = {};
        data.release_orders.forEach(d => roMap[d.date] = d.count);

        // Prepare chart data
        const chartData = {
            labels: sortedDates.map(d => frappe.datetime.str_to_user(d)),
            datasets: [
                {
                    name: 'Sales Orders',
                    values: sortedDates.map(d => soMap[d] || 0)
                },
                {
                    name: 'Release Orders',
                    values: sortedDates.map(d => roMap[d] || 0)
                },
                {
                    name: 'Delivery Notes',
                    values: sortedDates.map(d => dnMap[d] || 0)
                }
            ]
        };

        if (sortedDates.length === 0) {
            container.html('<div class="empty-state">No trend data available</div>');
            return;
        }

        new frappe.Chart(container[0], {
            data: chartData,
            type: 'line',
            height: 280,
            colors: ['#2563EB', '#0891B2', '#059669'],
            lineOptions: {
                regionFill: 1,
                hideDots: 0
            },
            axisOptions: {
                xAxisMode: 'tick',
                xIsSeries: true
            },
            tooltipOptions: {
                formatTooltipX: d => d,
                formatTooltipY: d => d
            }
        });
    }

    render_bales_stats(data) {
        // Status counts
        let packedImport = 0, packedInhouse = 0, dispatched = 0;

        data.by_status.forEach(item => {
            if (item.status === 'Packed Import') packedImport = item.count;
            else if (item.status === 'Packed In House') packedInhouse = item.count;
            else if (item.status === 'Dispatched') dispatched = item.count;
        });

        this.$wrapper.find('#bales-packed-import').text(this.format_number(packedImport));
        this.$wrapper.find('#bales-packed-inhouse').text(this.format_number(packedInhouse));
        this.$wrapper.find('#bales-dispatched').text(this.format_number(dispatched));
        this.$wrapper.find('#bales-period-dispatched').text(this.format_number(data.dispatched_in_period.count || 0));

        // Source chart
        const sourceContainer = this.$wrapper.find('#bales-source-chart');
        if (data.by_source && data.by_source.length > 0) {
            new frappe.Chart(sourceContainer[0], {
                data: {
                    labels: data.by_source.map(d => d.source || 'Unknown'),
                    datasets: [{
                        values: data.by_source.map(d => d.count)
                    }]
                },
                type: 'donut',
                height: 180,
                colors: ['#6366F1', '#8B5CF6', '#06B6D4', '#10B981']
            });
        } else {
            sourceContainer.html('<div class="empty-state">No source data</div>');
        }
    }

    render_batch_stats(data) {
        // Summary
        this.$wrapper.find('#batch-bundles').text(this.format_number(data.summary.total_bundles || 0));
        this.$wrapper.find('#batch-qty').text(this.format_number(Math.abs(data.summary.total_qty || 0)));
        this.$wrapper.find('#batch-dns').text(this.format_number(data.summary.unique_dns || 0));

        // Top items
        const container = this.$wrapper.find('#top-items-container');
        container.empty();

        if (!data.top_items || data.top_items.length === 0) {
            container.html('<div class="empty-state">No items data</div>');
            return;
        }

        data.top_items.forEach((item, index) => {
            const html = `
                <div class="top-item">
                    <div class="top-item-rank">${index + 1}</div>
                    <div class="top-item-info">
                        <div class="top-item-name">${frappe.utils.escape_html(item.item_name || item.item_code)}</div>
                        <div class="top-item-code">${frappe.utils.escape_html(item.item_code)}</div>
                    </div>
                    <div class="top-item-value">${this.format_number(item.total_consumed)}</div>
                </div>
            `;
            container.append(html);
        });
    }

    render_ro_stats(data) {
        // Status chart
        const chartContainer = this.$wrapper.find('#ro-status-chart');
        if (data.by_status && data.by_status.length > 0) {
            new frappe.Chart(chartContainer[0], {
                data: {
                    labels: data.by_status.map(d => d.status),
                    datasets: [{
                        values: data.by_status.map(d => d.count)
                    }]
                },
                type: 'bar',
                height: 150,
                colors: ['#94A3B8', '#6366F1', '#EF4444']
            });
        } else {
            chartContainer.html('<div class="empty-state">No RO data</div>');
        }

        // Conversion stats
        this.$wrapper.find('#ro-total').text(data.conversion.total_ro || 0);
        this.$wrapper.find('#ro-dn-created').text(data.conversion.dn_created || 0);

        // Recent ROs
        const container = this.$wrapper.find('#recent-ros-container');
        container.empty();

        if (!data.recent || data.recent.length === 0) {
            container.html('<div class="empty-state">No recent release orders</div>');
            return;
        }

        data.recent.forEach(ro => {
            const statusClass = ro.docstatus === 1 ? 'success' : 'draft';
            const html = `
                <div class="recent-item" data-doctype="Release Order" data-name="${ro.name}">
                    <div class="recent-item-info">
                        <div class="recent-item-name">${ro.name}</div>
                        <div class="recent-item-sub">${frappe.utils.escape_html(ro.customer_name || '')}</div>
                    </div>
                    <div class="recent-item-meta">
                        <div class="recent-item-value">${this.format_number(ro.total_qty)} Qty</div>
                        <div class="recent-item-date">${frappe.datetime.str_to_user(ro.posting_date)}</div>
                    </div>
                </div>
            `;
            container.append(html);
        });
    }

    render_top_customers(data) {
        const container = this.$wrapper.find('#top-customers-container');
        container.empty();

        if (!data || data.length === 0) {
            container.html('<div class="empty-state">No customer data available</div>');
            return;
        }

        data.forEach((customer, index) => {
            let rankClass = 'default';
            if (index === 0) rankClass = 'gold';
            else if (index === 1) rankClass = 'silver';
            else if (index === 2) rankClass = 'bronze';

            const html = `
                <div class="customer-item" data-customer="${customer.customer}">
                    <div class="customer-rank ${rankClass}">${index + 1}</div>
                    <div class="customer-info">
                        <div class="customer-name">${frappe.utils.escape_html(customer.customer_name || customer.customer)}</div>
                        <div class="customer-code">${frappe.utils.escape_html(customer.customer)}</div>
                    </div>
                    <div class="customer-stats">
                        <div class="customer-value">${this.format_currency(customer.total_value)}</div>
                        <div class="customer-orders">${customer.order_count} orders</div>
                    </div>
                </div>
            `;
            container.append(html);
        });
    }

    render_pending_documents(data) {
        // Pending Quotations
        this.render_pending_list(
            '#pending-quotations',
            '#pending-quotation-count',
            data.quotations,
            'Quotation',
            (item) => `
                <div class="pending-item-header">
                    <span class="pending-item-name">${item.name}</span>
                    <span class="pending-item-value">${this.format_currency(item.grand_total)}</span>
                </div>
                <div class="pending-item-sub">${frappe.utils.escape_html(item.customer_name || '')}</div>
            `
        );

        // Pending Sales Orders
        this.render_pending_list(
            '#pending-sos',
            '#pending-so-count',
            data.sales_orders,
            'Sales Order',
            (item) => `
                <div class="pending-item-header">
                    <span class="pending-item-name">${item.name}</span>
                    <span class="pending-item-value">${this.format_number(item.total_qty)} Qty</span>
                </div>
                <div class="pending-item-sub">${frappe.utils.escape_html(item.customer_name || '')} | ${frappe.datetime.str_to_user(item.transaction_date)}</div>
            `
        );

        // Draft Release Orders
        this.render_pending_list(
            '#pending-ros',
            '#pending-ro-count',
            data.release_orders,
            'Release Order',
            (item) => `
                <div class="pending-item-header">
                    <span class="pending-item-name">${item.name}</span>
                    <span class="pending-item-value">${this.format_number(item.total_qty)} Qty</span>
                </div>
                <div class="pending-item-sub">${frappe.utils.escape_html(item.customer_name || '')} | SO: ${item.sales_order || '-'}</div>
            `
        );

        // Draft Delivery Notes
        this.render_pending_list(
            '#pending-dns',
            '#pending-dn-count',
            data.delivery_notes,
            'Delivery Note',
            (item) => `
                <div class="pending-item-header">
                    <span class="pending-item-name">${item.name}</span>
                    <span class="pending-item-value">${this.format_number(item.total_qty)} Qty</span>
                </div>
                <div class="pending-item-sub">${frappe.utils.escape_html(item.customer_name || '')}</div>
            `
        );
    }

    render_pending_list(containerSelector, countSelector, items, doctype, templateFn) {
        const container = this.$wrapper.find(containerSelector);
        const countEl = this.$wrapper.find(countSelector);

        container.empty();
        countEl.text(items ? items.length : 0);

        if (!items || items.length === 0) {
            container.html('<div class="empty-state">No pending items</div>');
            return;
        }

        items.forEach(item => {
            const html = `
                <div class="pending-item" data-doctype="${doctype}" data-name="${item.name}">
                    ${templateFn(item)}
                </div>
            `;
            container.append(html);
        });
    }

    // Utility functions
    format_currency(value) {
        // Use simple formatting instead of frappe.format() which returns HTML
        const num = parseFloat(value) || 0;
        const formatted = num.toLocaleString('en-IN', {
            minimumFractionDigits: 2,
            maximumFractionDigits: 2
        });
        // Get currency symbol from system settings
        const symbol = frappe.boot.sysdefaults.currency || '₹';
        return symbol + ' ' + formatted;
    }

    format_number(value) {
        return (value || 0).toLocaleString('en-IN');
    }

    animate_number(selector, value) {
        const el = this.$wrapper.find(selector);
        const duration = 500;
        const start = 0;
        const increment = value / (duration / 16);
        let current = start;

        const animate = () => {
            current += increment;
            if (current < value) {
                el.text(Math.floor(current).toLocaleString('en-IN'));
                requestAnimationFrame(animate);
            } else {
                el.text(value.toLocaleString('en-IN'));
            }
        };

        if (value > 0) {
            animate();
        } else {
            el.text('0');
        }
    }
}

File: tcb_manufacturing_customizations/tcb_manufacturing_customizations/page/sales_flow_dashboard/sales_flow_dashboard.py
import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate, add_days, get_first_day, get_last_day


@frappe.whitelist()
def get_dashboard_data(from_date=None, to_date=None, company=None):
    """Get all dashboard data in a single call"""
    if not from_date:
        from_date = get_first_day(nowdate())
    if not to_date:
        to_date = get_last_day(nowdate())
    if not company:
        company = frappe.defaults.get_user_default("Company")

    filters = {
        "from_date": from_date,
        "to_date": to_date,
        "company": company,
    }

    return {
        "summary_cards": get_summary_cards(filters),
        "flow_funnel": get_flow_funnel_data(filters),
        "daily_trend": get_daily_trend(filters),
        "release_order_stats": get_release_order_stats(filters),
        "bales_stats": get_bales_stats(filters),
        "batch_consumption": get_batch_consumption_stats(filters),
        "top_customers": get_top_customers(filters),
        "pending_documents": get_pending_documents(filters),
    }


def get_summary_cards(filters):
    """Get summary card data for Quotation → SO → RO → DN flow"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    # Quotations
    quotation_data = frappe.db.sql(
        """
        SELECT
            COUNT(*) as count,
            SUM(grand_total) as total_value,
            SUM(CASE WHEN status = 'Open' THEN 1 ELSE 0 END) as open_count,
            SUM(CASE WHEN status = 'Ordered' THEN 1 ELSE 0 END) as ordered_count
        FROM `tabQuotation`
        WHERE docstatus = 1
            AND transaction_date BETWEEN %s AND %s
            AND company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    # Sales Orders
    so_data = frappe.db.sql(
        """
        SELECT
            COUNT(*) as count,
            SUM(grand_total) as total_value,
            SUM(total_qty) as total_qty,
            SUM(CASE WHEN status IN ('To Deliver', 'To Deliver and Bill') THEN 1 ELSE 0 END) as pending_delivery
        FROM `tabSales Order`
        WHERE docstatus = 1
            AND transaction_date BETWEEN %s AND %s
            AND company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    # Release Orders
    ro_data = frappe.db.sql(
        """
        SELECT
            COUNT(*) as count,
            SUM(total_qty) as total_qty,
            SUM(CASE WHEN docstatus = 0 THEN 1 ELSE 0 END) as draft_count,
            SUM(CASE WHEN docstatus = 1 THEN 1 ELSE 0 END) as submitted_count
        FROM `tabRelease Order`
        WHERE posting_date BETWEEN %s AND %s
            AND company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    # Delivery Notes
    dn_data = frappe.db.sql(
        """
        SELECT
            COUNT(*) as count,
            SUM(grand_total) as total_value,
            SUM(total_qty) as total_qty
        FROM `tabDelivery Note`
        WHERE docstatus = 1
            AND posting_date BETWEEN %s AND %s
            AND company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    return {
        "quotation": {
            "count": quotation_data.get("count") or 0,
            "value": flt(quotation_data.get("total_value")),
            "open": quotation_data.get("open_count") or 0,
            "ordered": quotation_data.get("ordered_count") or 0,
        },
        "sales_order": {
            "count": so_data.get("count") or 0,
            "value": flt(so_data.get("total_value")),
            "qty": flt(so_data.get("total_qty")),
            "pending_delivery": so_data.get("pending_delivery") or 0,
        },
        "release_order": {
            "count": ro_data.get("count") or 0,
            "qty": flt(ro_data.get("total_qty")),
            "draft": ro_data.get("draft_count") or 0,
            "submitted": ro_data.get("submitted_count") or 0,
        },
        "delivery_note": {
            "count": dn_data.get("count") or 0,
            "value": flt(dn_data.get("total_value")),
            "qty": flt(dn_data.get("total_qty")),
        },
    }


def get_flow_funnel_data(filters):
    """Get funnel data showing conversion from Quotation to DN"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    # Quotations created
    quotations = frappe.db.count(
        "Quotation",
        {
            "docstatus": 1,
            "transaction_date": ["between", [from_date, to_date]],
            "company": company,
        },
    )

    # Quotations converted to SO
    converted_quotations = frappe.db.sql(
        """
        SELECT COUNT(DISTINCT q.name)
        FROM `tabQuotation` q
        INNER JOIN `tabSales Order Item` soi ON soi.prevdoc_docname = q.name
        INNER JOIN `tabSales Order` so ON so.name = soi.parent
        WHERE q.docstatus = 1
            AND so.docstatus = 1
            AND q.transaction_date BETWEEN %s AND %s
            AND q.company = %s
    """,
        (from_date, to_date, company),
    )[0][0] or 0

    # Sales Orders
    sales_orders = frappe.db.count(
        "Sales Order",
        {
            "docstatus": 1,
            "transaction_date": ["between", [from_date, to_date]],
            "company": company,
        },
    )

    # Release Orders
    release_orders = frappe.db.count(
        "Release Order",
        {
            "docstatus": 1,
            "posting_date": ["between", [from_date, to_date]],
            "company": company,
        },
    )

    # Delivery Notes
    delivery_notes = frappe.db.count(
        "Delivery Note",
        {
            "docstatus": 1,
            "posting_date": ["between", [from_date, to_date]],
            "company": company,
        },
    )

    return [
        {"stage": "Quotations", "count": quotations, "color": "#7C3AED"},
        {"stage": "Converted to SO", "count": converted_quotations, "color": "#8B5CF6"},
        {"stage": "Sales Orders", "count": sales_orders, "color": "#2563EB"},
        {"stage": "Release Orders", "count": release_orders, "color": "#0891B2"},
        {"stage": "Delivery Notes", "count": delivery_notes, "color": "#059669"},
    ]


def get_daily_trend(filters):
    """Get daily document creation trend"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    # Get daily SO count
    so_trend = frappe.db.sql(
        """
        SELECT DATE(transaction_date) as date, COUNT(*) as count, SUM(total_qty) as qty
        FROM `tabSales Order`
        WHERE docstatus = 1
            AND transaction_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY DATE(transaction_date)
        ORDER BY date
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    # Get daily DN count
    dn_trend = frappe.db.sql(
        """
        SELECT DATE(posting_date) as date, COUNT(*) as count, SUM(total_qty) as qty
        FROM `tabDelivery Note`
        WHERE docstatus = 1
            AND posting_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY DATE(posting_date)
        ORDER BY date
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    # Get daily RO count
    ro_trend = frappe.db.sql(
        """
        SELECT DATE(posting_date) as date, COUNT(*) as count, SUM(total_qty) as qty
        FROM `tabRelease Order`
        WHERE docstatus = 1
            AND posting_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY DATE(posting_date)
        ORDER BY date
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    return {
        "sales_orders": so_trend,
        "delivery_notes": dn_trend,
        "release_orders": ro_trend,
    }


def get_release_order_stats(filters):
    """Get Release Order specific statistics"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    # Release Order by status
    status_data = frappe.db.sql(
        """
        SELECT
            CASE
                WHEN docstatus = 0 THEN 'Draft'
                WHEN docstatus = 1 THEN 'Submitted'
                WHEN docstatus = 2 THEN 'Cancelled'
            END as status,
            COUNT(*) as count,
            SUM(total_qty) as qty
        FROM `tabRelease Order`
        WHERE posting_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY docstatus
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    # RO to DN conversion stats
    conversion_stats = frappe.db.sql(
        """
        SELECT
            COUNT(DISTINCT ro.name) as total_ro,
            COUNT(DISTINCT dn.name) as dn_created
        FROM `tabRelease Order` ro
        LEFT JOIN `tabDelivery Note` dn ON dn.custom_release_order = ro.name AND dn.docstatus = 1
        WHERE ro.docstatus = 1
            AND ro.posting_date BETWEEN %s AND %s
            AND ro.company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    # Recent Release Orders
    recent_ros = frappe.db.sql(
        """
        SELECT
            ro.name,
            ro.sales_order,
            ro.customer_name,
            ro.total_qty,
            ro.posting_date,
            ro.docstatus
        FROM `tabRelease Order` ro
        WHERE ro.posting_date BETWEEN %s AND %s
            AND ro.company = %s
        ORDER BY ro.creation DESC
        LIMIT 5
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    return {
        "by_status": status_data,
        "conversion": conversion_stats,
        "recent": recent_ros,
    }


def get_bales_stats(filters):
    """Get Bales related statistics"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")

    # Bales by status
    status_data = frappe.db.sql(
        """
        SELECT
            bales_status as status,
            COUNT(*) as count,
            SUM(bale_qty) as qty
        FROM `tabBales`
        WHERE docstatus = 1
        GROUP BY bales_status
    """,
        as_dict=True,
    )

    # Bales dispatched in period
    dispatched = frappe.db.sql(
        """
        SELECT
            COUNT(*) as count,
            SUM(bale_qty) as qty
        FROM `tabBales`
        WHERE docstatus = 1
            AND bales_status = 'Dispatched'
            AND modified BETWEEN %s AND %s
    """,
        (from_date, to_date),
        as_dict=True,
    )[0]

    # Bales by source type
    by_source = frappe.db.sql(
        """
        SELECT
            source_document_type as source,
            COUNT(*) as count,
            SUM(bale_qty) as qty
        FROM `tabBales`
        WHERE docstatus = 1
        GROUP BY source_document_type
    """,
        as_dict=True,
    )

    # Recent bales linked to DNs
    recent_dn_bales = frappe.db.sql(
        """
        SELECT
            dnb.delivery_note,
            dnb.total_bales,
            dnb.total_qty,
            dn.customer_name,
            dn.posting_date
        FROM `tabDelivery Note Bales` dnb
        INNER JOIN `tabDelivery Note` dn ON dn.name = dnb.delivery_note
        WHERE dn.posting_date BETWEEN %s AND %s
        ORDER BY dnb.creation DESC
        LIMIT 5
    """,
        (from_date, to_date),
        as_dict=True,
    )

    return {
        "by_status": status_data,
        "dispatched_in_period": dispatched,
        "by_source": by_source,
        "recent_dn_bales": recent_dn_bales,
    }


def get_batch_consumption_stats(filters):
    """Get Serial and Batch Bundle consumption statistics"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    # Bundles created for Delivery Notes
    bundle_stats = frappe.db.sql(
        """
        SELECT
            COUNT(*) as total_bundles,
            SUM(total_qty) as total_qty,
            COUNT(DISTINCT voucher_no) as unique_dns
        FROM `tabSerial and Batch Bundle`
        WHERE voucher_type = 'Delivery Note'
            AND type_of_transaction = 'Outward'
            AND docstatus = 1
            AND posting_date BETWEEN %s AND %s
            AND company = %s
    """,
        (from_date, to_date, company),
        as_dict=True,
    )[0]

    # Top items by batch consumption
    top_items = frappe.db.sql(
        """
        SELECT
            sbb.item_code,
            i.item_name,
            COUNT(*) as bundle_count,
            SUM(ABS(sbb.total_qty)) as total_consumed
        FROM `tabSerial and Batch Bundle` sbb
        INNER JOIN `tabItem` i ON i.name = sbb.item_code
        WHERE sbb.voucher_type = 'Delivery Note'
            AND sbb.type_of_transaction = 'Outward'
            AND sbb.docstatus = 1
            AND sbb.posting_date BETWEEN %s AND %s
            AND sbb.company = %s
        GROUP BY sbb.item_code
        ORDER BY total_consumed DESC
        LIMIT 5
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    # Daily batch consumption trend
    daily_consumption = frappe.db.sql(
        """
        SELECT
            DATE(posting_date) as date,
            COUNT(*) as bundles,
            SUM(ABS(total_qty)) as qty
        FROM `tabSerial and Batch Bundle`
        WHERE voucher_type = 'Delivery Note'
            AND type_of_transaction = 'Outward'
            AND docstatus = 1
            AND posting_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY DATE(posting_date)
        ORDER BY date
    """,
        (from_date, to_date, company),
        as_dict=True,
    )

    return {
        "summary": bundle_stats,
        "top_items": top_items,
        "daily_trend": daily_consumption,
    }


def get_top_customers(filters):
    """Get top customers by sales value"""
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    company = filters.get("company")

    return frappe.db.sql(
        """
        SELECT
            customer,
            customer_name,
            COUNT(*) as order_count,
            SUM(grand_total) as total_value,
            SUM(total_qty) as total_qty
        FROM `tabSales Order`
        WHERE docstatus = 1
            AND transaction_date BETWEEN %s AND %s
            AND company = %s
        GROUP BY customer
        ORDER BY total_value DESC
        LIMIT 5
    """,
        (from_date, to_date, company),
        as_dict=True,
    )


def get_pending_documents(filters):
    """Get pending documents requiring action"""
    company = filters.get("company")

    # Pending Quotations (Open)
    pending_quotations = frappe.db.sql(
        """
        SELECT name, customer_name, grand_total, transaction_date
        FROM `tabQuotation`
        WHERE docstatus = 1
            AND status = 'Open'
            AND company = %s
        ORDER BY transaction_date DESC
        LIMIT 5
    """,
        (company,),
        as_dict=True,
    )

    # Pending SO (To Deliver)
    pending_so = frappe.db.sql(
        """
        SELECT name, customer_name, grand_total, total_qty, transaction_date
        FROM `tabSales Order`
        WHERE docstatus = 1
            AND status IN ('To Deliver', 'To Deliver and Bill')
            AND company = %s
        ORDER BY transaction_date ASC
        LIMIT 5
    """,
        (company,),
        as_dict=True,
    )

    # Draft Release Orders
    draft_ro = frappe.db.sql(
        """
        SELECT name, sales_order, customer_name, total_qty, posting_date
        FROM `tabRelease Order`
        WHERE docstatus = 0
            AND company = %s
        ORDER BY creation DESC
        LIMIT 5
    """,
        (company,),
        as_dict=True,
    )

    # Draft Delivery Notes
    draft_dn = frappe.db.sql(
        """
        SELECT name, customer_name, grand_total, total_qty, posting_date
        FROM `tabDelivery Note`
        WHERE docstatus = 0
            AND company = %s
        ORDER BY creation DESC
        LIMIT 5
    """,
        (company,),
        as_dict=True,
    )

    return {
        "quotations": pending_quotations,
        "sales_orders": pending_so,
        "release_orders": draft_ro,
        "delivery_notes": draft_dn,
    }
